corehq/apps/users/user_data.py

```diff
@@ -174,8 +180,8 @@ def get(self, key, default=None):
         return self.to_dict().get(key, default)
 
     def __setitem__(self, key, value):
-        if key in self._provided_by_system:
-            if value == self._provided_by_system[key]:
+        if key in self._keys_provided_by_system:
+            if key in self._provided_by_system and value == self._provided_by_system[key]:
```

Comment:
Maybe consider renaming `_keys_provided_by_system` to something more distinct from `_provided_by_system`, to make this easier to read, possibly `_system_keys`. For me, it was hard to figure out what these two lines actually mean without reading through them a few times. Not sure if you could also use `get` or `getattr` here but that's another approach that could help with clarity, along the lines of:
```python
if key in self._system_keys:
    if value == self._provided_by_system.get(key):
```
Unless we're concerned with someone trying to pass in a null value where one already exists, which would be interesting.

Reply:
Any value that is provided by system shouldn't be assigned in this set method. So even null value should be passed to it. But I will make changes to the function name!

Reply:
If we use `''` instead of `None` in getattr(self._couch_user, 'location_id', ''):, do we still need to user fallback like `object()` ?

Reply:
Not sure, because setting it to `''` is also a real change, like `None`.  I'd expect both of these to fail:
```python
user_data['commcare_project'] = None
user_data['commcare_project'] = ''
```
Nice thing about `object()` is there's nothing you can pass in that will equal it.  You could also use a string like "MISSING", though that's a little messier.  `...` is also sometimes used for this sort of thing, as a placeholder meaning "not provided".

Reply:
Thank you! Fixed in: afdbd8d89c5a635e5214d4f2070027b42bbbf13e